### PR TITLE
Probe utility methods

### PR DESCRIPTION
**Pull request by Conorbro**
Utility methods for probes for #154

### CODE REVIEW

File: utils_test.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+	dto "github.com/prometheus/client_model/go"
+)
+
+// Check if expected results are in the registry
+func checkRegistryResults(expRes map[string]float64, mfs []*dto.MetricFamily, t *testing.T) {
+	if len(mfs) == 0 {
+		t.Fatal("No metrics returned in registry")
+	}
+	for i := range mfs {
+		if _, ok := expRes[mfs[i].GetName()]; ok {
```
Comment (brian-brazil) [inline utils_test.go]: You've still got the problem where if mfs was empty that this check would succeed

File: utils_test.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+// Check if expected results are in the registry
+func checkRegistryResults(expRes map[string]float64, mfs []*dto.MetricFamily, t *testing.T) {
+
+	res := make(map[string]float64)
+
+	if len(mfs) == 0 {
```
Comment (brian-brazil) [inline utils_test.go]: No need to special care this

File: utils_test.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import (
+	"fmt"
+	"testing"
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+// Check if expected results are in the registry
+func checkRegistryResults(expRes map[string]float64, mfs []*dto.MetricFamily, t *testing.T) {
+
```
Comment (brian-brazil) [inline utils_test.go]: You can reduce the vertical whitespace

File: utils_test.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+		t.Fatal(fmt.Errorf("No metrics returned in registry"))
+	}
+
+	for i := range mfs {
+		res[mfs[i].GetName()] = mfs[i].Metric[0].GetGauge().GetValue()
+	}
+
+	for k, v := range expRes {
+		if _, ok := res[k]; !ok || res[k] != v {
+			t.Fatal(fmt.Errorf("Expected: %v: %v, got: %v: %v", k, v, k, res[k]))
```
Comment (brian-brazil) [inline utils_test.go]: `res[k]` here can cause an exception if it's missing, use a variable created in the if

File: main.go [hunk 5]
```diff
@@ -53,7 +53,7 @@ type Module struct {
 }
 
 type HTTPProbe struct {
-	// Defaults to 2xx.
```
Comment (brian-brazil) [inline main.go]: This comment should stay here.

File: main.go [hunk 6]
```diff
@@ -87,22 +84,22 @@ type ICMPProbe struct {
 }
 
 type DNSProbe struct {
-	Protocol            string         `yaml:"protocol"` // Defaults to "udp".
+	PreferredIPProtocol string         `yaml:"preferred_ip_protocol"`
+	Protocol            string         `yaml:"transport_protocol"`
```
Comment (brian-brazil) [inline main.go]: TransportProtocol would be clearer

File: http.go [hunk 7]
```diff
@@ -100,46 +100,42 @@ func probeHTTP(target string, w http.ResponseWriter, module Module, registry *pr
...
 	}
+
 	if !strings.HasPrefix(target, "http://") && !strings.HasPrefix(target, "https://") {
 		target = "http://" + target
 	}
 
-	dialProtocol = module.HTTP.Protocol
-	if module.HTTP.Protocol == "tcp" {
-		targetURL, err := url.Parse(target)
+	dialProtocol = "tcp"
```
Comment (brian-brazil) [inline http.go]: This value is never used

File: dns.go [hunk 8]
```diff
@@ -113,11 +112,11 @@ func probeDNS(target string, w http.ResponseWriter, module Module, registry *pro
...
 	}()
 
-	if module.DNS.Protocol == "" {
-		module.DNS.Protocol = "udp"
+	if module.DNS.TransportProtocol == "" {
+		module.DNS.TransportProtocol = "udp"
 	}
 
-	if (module.DNS.Protocol == "tcp" || module.DNS.Protocol == "udp") && module.DNS.PreferredIPProtocol == "" {
+	if (module.DNS.TransportProtocol == "tcp" || module.DNS.TransportProtocol == "udp") && module.DNS.PreferredIPProtocol == "" {
```
Comment (brian-brazil) [inline dns.go]: TransportProtocol is always one of these two

File: dns.go [hunk 9]
```diff
@@ -126,8 +125,8 @@ func probeDNS(target string, w http.ResponseWriter, module Module, registry *pro
 		fallbackProtocol = "ip6"
 	}
 
-	dialProtocol = module.DNS.Protocol
-	if module.DNS.Protocol == "udp" || module.DNS.Protocol == "tcp" {
+	dialProtocol = module.DNS.TransportProtocol
```
Comment (brian-brazil) [inline dns.go]: This value is not used before being set below

File: http.go [hunk 10]
```diff
@@ -100,46 +100,41 @@ func probeHTTP(target string, w http.ResponseWriter, module Module, registry *pr
...
 
-	dialProtocol = module.HTTP.Protocol
-	if module.HTTP.Protocol == "tcp" {
-		targetURL, err := url.Parse(target)
+	targetURL, err := url.Parse(target)
+	if err != nil {
+		return false
+	}
+	targetHost, _, err := net.SplitHostPort(targetURL.Host)
+	// If split fails, assuming it's a hostname without port part
```
Comment (brian-brazil) [inline http.go]: comments should end in a full stop

File: main.go [hunk 11]
```diff
@@ -54,6 +54,7 @@ type Module struct {
 
 type HTTPProbe struct {
 	// Defaults to 2xx.
```
Comment (brian-brazil) [inline main.go]: This comment should be one line down.

**Comment by Conorbro** [comment]
@brian-brazil

File: utils.go [hunk 12]
```diff
@@ -0,0 +1,64 @@
+package main
+
+import (
+	"net"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Returns the preferedIPProtocol, the dialProtocol, and sets the probeIPProtocolGauge.
+func configureProtocol(preferredIPProtocol string, targetAddress string, module Module, probeIPProtocolGauge prometheus.Gauge) (string, string, error) {
```
Comment (brian-brazil) [inline utils.go]: You don't need to know the module

Also, create the Gauge in this function. Take in the registry.

File: utils.go [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+		if err != nil {
+			ip, err = net.ResolveIPAddr(fallbackProtocol, targetAddress)
+			if err != nil {
+				return "", "", err
+			}
+		}
+	}
+
+	if ip == nil {
+		if module.DNS.TransportProtocol[len(module.DNS.TransportProtocol)-1] == '6' {
```
Comment (brian-brazil) [inline utils.go]: There should be no mention of DNS here

File: utils.go [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+		} else {
+			probeIPProtocolGauge.Set(4)
+			dialProtocol = "tcp4"
+		}
+		return preferredIPProtocol, dialProtocol, nil
+	}
+
+	if ip.IP.To4() == nil {
+		probeIPProtocolGauge.Set(6)
+		dialProtocol = module.DNS.TransportProtocol + "6"
```
Comment (brian-brazil) [inline utils.go]: This is DNS specific logic, it should remain in the dns function

dialProtocol doesn't belong in here.

File: utils.go [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+	if err != nil {
+		ip, err = net.ResolveIPAddr(fallbackProtocol, targetAddress)
+		if err != nil {
+			return "", "", err
+		}
+	}
+
+	if ip.IP.To4() == nil {
+		probeIPProtocolGauge.Set(6)
+		dialProtocol = "tcp6"
```
Comment (brian-brazil) [inline utils.go]: Handle the dial protocol in each of the probes, it's not always tcp.

File: utils.go [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+	ip, err := net.ResolveIPAddr(preferredIPProtocol, targetAddress)
+	if err != nil {
+		ip, err = net.ResolveIPAddr(fallbackProtocol, targetAddress)
+		if err != nil {
+			return "", "", err
+		}
+	}
+
+	if ip.IP.To4() == nil {
```
Comment (brian-brazil) [inline utils.go]: The IP should be returned, otherwise it'll be reresolved.

File: utils.go [hunk 17]
```diff
@@ -0,0 +1,47 @@
+package main
+
+import (
+	"net"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Returns the preferedIPProtocol, the dialProtocol, and sets the probeIPProtocolGauge.
+func configureProtocol(preferredIPProtocol string, targetAddress string, registry *prometheus.Registry) (string, string, error) {
```
Comment (brian-brazil) [inline utils.go]: choose would be clearer than configure, as this function is stateless.

File: utils.go [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+
+	if preferredIPProtocol == "ip6" || preferredIPProtocol == "" {
+		preferredIPProtocol = "ip6"
+		fallbackProtocol = "ip4"
+	} else {
+		preferredIPProtocol = "ip4"
+		fallbackProtocol = "ip6"
+	}
+
+	if transportProtocol == "icmp4" {
```
Comment (brian-brazil) [inline utils.go]: As we discussed, this function should only be looking at the network protocol - not the transport protocol. Have the transport protocol just be "tcp" without the network protocol mentioned.

File: dns.go [hunk 19]
```diff
@@ -113,43 +107,33 @@ func probeDNS(target string, w http.ResponseWriter, module Module, registry *pro
...
-		targetAddress, _, _ := net.SplitHostPort(target)
-		ip, err := net.ResolveIPAddr(module.DNS.PreferredIPProtocol, targetAddress)
-		if err != nil {
-			ip, err = net.ResolveIPAddr(fallbackProtocol, targetAddress)
-			if err != nil {
-				return false
-			}
-		}
-
+	if module.DNS.IPProtocol == "" {
```
Comment (brian-brazil) [inline dns.go]: You don't need this, just the below code covers it

File: dns.go [hunk 20]
```diff
@@ -113,43 +107,33 @@ func probeDNS(target string, w http.ResponseWriter, module Module, registry *pro
...
+		module.DNS.TransportProtocol = "udp"
 	}
-	if module.DNS.PreferredIPProtocol == "ip6" {
-		fallbackProtocol = "ip4"
+
+	dialProtocol = module.DNS.TransportProtocol
+	if module.DNS.TransportProtocol == "udp" || module.DNS.TransportProtocol == "tcp" {
+		targetAddr, _, _ := net.SplitHostPort(target)
+		module.DNS.PreferredIPProtocol, ip, err = chooseProtocol(module.DNS.PreferredIPProtocol, targetAddr, registry)
 	} else {
```
Comment (brian-brazil) [inline dns.go]: This case should cause an error, as there's only two valid protocols.

File: icmp.go [hunk 21]
```diff
@@ -41,54 +40,37 @@ func getICMPSequence() uint16 {
...
-		module.ICMP.Protocol = "icmp"
-	}
+	var ip *net.IPAddr
+	var err error
 
-	// In case of ICMP prefer IPv6 by default
-	if module.ICMP.Protocol == "icmp" && module.ICMP.PreferredIPProtocol == "" {
-		module.ICMP.PreferredIPProtocol = "ip6"
+	// Defaults to IPv4 to be compatible with older versions
+	if module.ICMP.TransportProtocol == "" {
```
Comment (brian-brazil) [inline icmp.go]: ICMP is always ICMP, there's no need for this to be configurable.

File: README.md [hunk 22]
```diff
@@ -45,7 +45,7 @@ modules:
         - "Download the latest version here"
       tls_config:
         insecure_skip_verify: false
-      protocol: "tcp" # accepts "tcp/tcp4/tcp6", defaults to "tcp"
+      protocol: "tcp" # accepts "tcp", defaults to "tcp"
```
Comment (brian-brazil) [inline README.md]: remove this, it's always tcp

File: README.md [hunk 23]
```diff
@@ -45,7 +45,7 @@ modules:
...
       preferred_ip_protocol: "ip4" # used for "tcp", defaults to "ip6"
```
Comment (brian-brazil) [inline README.md]: this needs updating

File: README.md [hunk 24]
```diff
@@ -59,7 +59,7 @@ modules:
     prober: tcp
     timeout: 5s
     tcp:
-      protocol: "tcp4"
+      protocol: "tcp"
```
Comment (brian-brazil) [inline README.md]: remove this, it's always tcp

File: dns.go [hunk 25]
```diff
@@ -113,41 +107,27 @@ func probeDNS(target string, w http.ResponseWriter, module Module, registry *pro
...
-			}
-		}
-
-		if ip.IP.To4() == nil {
-			dialProtocol = module.DNS.Protocol + "6"
-		} else {
-			dialProtocol = module.DNS.Protocol + "4"
-		}
+	dialProtocol = module.DNS.TransportProtocol
+	if module.DNS.TransportProtocol == "udp" || module.DNS.TransportProtocol == "tcp" {
```
Comment (brian-brazil) [inline dns.go]: otherwise it's a config error

File: dns.go [hunk 26]
```diff
@@ -113,41 +107,27 @@ func probeDNS(target string, w http.ResponseWriter, module Module, registry *pro
...
+		targetAddr, _, _ := net.SplitHostPort(target)
+		module.DNS.PreferredIPProtocol, ip, err = chooseProtocol(module.DNS.PreferredIPProtocol, targetAddr, registry)
+	}
+	if err != nil {
```
Comment (brian-brazil) [inline dns.go]: This would make more sense inside the if

File: dns.go [hunk 27]
```diff
@@ -113,41 +107,27 @@ func probeDNS(target string, w http.ResponseWriter, module Module, registry *pro
...
-
-		if ip.IP.To4() == nil {
-			dialProtocol = module.DNS.Protocol + "6"
-		} else {
-			dialProtocol = module.DNS.Protocol + "4"
-		}
+	dialProtocol = module.DNS.TransportProtocol
+	if module.DNS.TransportProtocol == "udp" || module.DNS.TransportProtocol == "tcp" {
+		targetAddr, _, _ := net.SplitHostPort(target)
+		module.DNS.PreferredIPProtocol, ip, err = chooseProtocol(module.DNS.PreferredIPProtocol, targetAddr, registry)
```
Comment (brian-brazil) [inline dns.go]: use a local variable, don't change the config

File: http.go [hunk 28]
```diff
@@ -96,56 +91,34 @@ func probeHTTP(target string, w http.ResponseWriter, module Module, registry *pr
...
-			dialProtocol = "tcp4"
-		}
+	module.HTTP.PreferredIPProtocol, ip, err = chooseProtocol(module.HTTP.PreferredIPProtocol, targetHost, registry)
+	if err != nil {
+		return false
 	}
 
-	if dialProtocol == "tcp6" {
-		probeIPProtocolGauge.Set(6)
+	if ip.IP.To4() == nil {
```
Comment (brian-brazil) [inline http.go]: We should be using this ip address, not reresolving when dialing.

File: icmp.go [hunk 29]
```diff
@@ -41,54 +40,32 @@ func getICMPSequence() uint16 {
...
-		fallbackProtocol = "ip4"
-	} else {
-		fallbackProtocol = "ip6"
+	module.ICMP.PreferredIPProtocol, ip, err = chooseProtocol(module.ICMP.PreferredIPProtocol, target, registry)
+	if err != nil {
+		log.Error(err)
+		return false
 	}
 
 	resolveStart := time.Now()
```
Comment (brian-brazil) [inline icmp.go]: This logic should be inside the utility function.

File: main.go [hunk 30]
```diff
@@ -74,35 +73,35 @@ type QueryResponse struct {
...
 	QueryResponse       []QueryResponse  `yaml:"query_response"`
 	TLS                 bool             `yaml:"tls"`
 	TLSConfig           config.TLSConfig `yaml:"tls_config"`
-	Protocol            string           `yaml:"protocol"`              // Defaults to "tcp".
-	PreferredIPProtocol string           `yaml:"preferred_ip_protocol"` // Defaults to "ip6".
 }
 
 type ICMPProbe struct {
-	Protocol            string `yaml:"protocol"`              // Defaults to "icmp4".
+	TransportProtocol   string `yaml:"transport_protocol"`    // Defaults to "icmp4".
```
Comment (brian-brazil) [inline main.go]: remove this, it's not needed

File: main.go [hunk 31]
```diff
@@ -74,35 +73,35 @@ type QueryResponse struct {
...
 	PreferredIPProtocol string `yaml:"preferred_ip_protocol"` // Defaults to "ip6".
 }
 
 type DNSProbe struct {
-	Protocol            string         `yaml:"protocol"` // Defaults to "udp".
+	PreferredIPProtocol string         `yaml:"preferred_ip_protocol"`
+	IPProtocol          string         `yaml:"ip_protocol"`
```
Comment (brian-brazil) [inline main.go]: this is unused, remove it

File: README.md [hunk 32]
```diff
@@ -58,8 +57,6 @@ modules:
   tcp_connect_v4_example:
```
Comment (brian-brazil) [inline README.md]: This example no longer does v4

File: README.md [hunk 33]
```diff
@@ -98,7 +95,7 @@ modules:
   dns_tcp_example:
     prober: dns
     dns:
-      protocol: "tcp" # accepts "tcp/tcp4/tcp6/udp/udp4/udp6", defaults to "udp"
+      protocol: "tcp" # accepts "tcp/udp", defaults to "udp"
       preferred_ip_protocol: "ip4" # used for "udp/tcp", defaults to "ip6"
```
Comment (brian-brazil) [inline README.md]: no need to mention tcp/udp here

File: dns.go [hunk 34]
```diff
@@ -113,41 +107,28 @@ func probeDNS(target string, w http.ResponseWriter, module Module, registry *pro
...
-		}
-
-		if ip.IP.To4() == nil {
-			dialProtocol = module.DNS.Protocol + "6"
-		} else {
-			dialProtocol = module.DNS.Protocol + "4"
+			log.Error(err)
+			return false
 		}
+	} else {
```
Comment (brian-brazil) [inline dns.go]: This should return false

File: http.go [hunk 35]
```diff
@@ -96,56 +91,32 @@ func probeHTTP(target string, w http.ResponseWriter, module Module, registry *pr
...
-			dialProtocol = "tcp4"
-		}
+	ip, err := chooseProtocol(module.HTTP.PreferredIPProtocol, targetHost, registry)
+	if err != nil {
+		return false
 	}
 
-	if dialProtocol == "tcp6" {
-		probeIPProtocolGauge.Set(6)
+	if ip == nil {
```
Comment (brian-brazil) [inline http.go]: Shouldn't this be checking the type of the ip? This is also not needed, the IP alone implies the protocol.

File: http.go [hunk 36]
```diff
@@ -180,6 +151,8 @@ func probeHTTP(target string, w http.ResponseWriter, module Module, registry *pr
 	}
 
 	request, err := http.NewRequest(config.Method, target, nil)
+	request.Header.Add("HOST", request.Host)
```
Comment (brian-brazil) [inline http.go]: "Host" is more common

File: http.go [hunk 37]
```diff
@@ -180,6 +151,8 @@ func probeHTTP(target string, w http.ResponseWriter, module Module, registry *pr
...
+	request.Host = ip.String()
```
Comment (brian-brazil) [inline http.go]: You can just set this to the URL's original host, it'll make the header do the right thing.

You then need to change the target URL to have the IP

File: utils.go [hunk 38]
```diff
@@ -0,0 +1,62 @@
...
+	var fallbackProtocol string
+
+	probeDNSLookupTimeSeconds := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "probe_dns_lookup_time_seconds",
+		Help: "Returns the time taken for probe dns lookup in seconds",
+	})
+
+	probeIPProtocolGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "probe_ip_protocol",
+		Help: "Specifies whether probe ip protocl is IP4 or IP6",
```
Comment (brian-brazil) [inline utils.go]: protocol

File: utils.go [hunk 39]
```diff
@@ -0,0 +1,62 @@
...
+	})
+	registry.MustRegister(probeIPProtocolGauge)
+	registry.MustRegister(probeDNSLookupTimeSeconds)
+
+	resolveStart := time.Now()
+	probeDNSLookupTimeSeconds.Add(time.Since(resolveStart).Seconds())
```
Comment (brian-brazil) [inline utils.go]: Shouldn't this be after the resolution?

File: README.md [hunk 40]
```diff
@@ -98,7 +95,7 @@ modules:
   dns_tcp_example:
     prober: dns
     dns:
-      protocol: "tcp" # accepts "tcp/tcp4/tcp6/udp/udp4/udp6", defaults to "udp"
+      protocol: "tcp" # defaults to "udp"
       preferred_ip_protocol: "ip4" # used for "udp/tcp", defaults to "ip6"
```
Comment (brian-brazil) [inline README.md]: no need to mention udp/tcp

File: README.md [hunk 41]
```diff
@@ -55,11 +54,9 @@ modules:
       headers:
         Content-Type: application/json
       body: '{}'
-  tcp_connect_v4_example:
+  tcp_connect_example:
```
Comment (brian-brazil) [inline README.md]: Below in icmp_example, there shouldn't be a protocol.

File: http.go [hunk 42]
```diff
@@ -180,6 +151,10 @@ func probeHTTP(target string, w http.ResponseWriter, module Module, registry *pr
 	}
 
 	request, err := http.NewRequest(config.Method, target, nil)
+	request.Header.Add("Host", request.Host)
+	request.Host = targetURL.Host
+	targetURL.Host = ip.String() + targetPort
```
Comment (brian-brazil) [inline http.go]: What if there's no port? You've also missing the colon.

File: http.go [hunk 43]
```diff
@@ -180,6 +151,10 @@ func probeHTTP(target string, w http.ResponseWriter, module Module, registry *pr
 	}
 
 	request, err := http.NewRequest(config.Method, target, nil)
+	request.Header.Add("Host", request.Host)
```
Comment (brian-brazil) [inline http.go]: You don't need this given the next line.

File: tcp.go [hunk 44]
```diff
@@ -18,52 +18,31 @@ import (
...
-			dialProtocol = "tcp4"
-		}
+	ip, err := chooseProtocol(module.TCP.PreferredIPProtocol, targetAddress, registry)
+	if err != nil {
+		return nil, err
 	}
 
-	if dialProtocol == "tcp6" {
-		protocolProbeGauge.Set(6)
+	if ip.IP.To4() == nil {
```
Comment (brian-brazil) [inline tcp.go]: You need to use the IP below

File: tcp.go [hunk 45]
```diff
@@ -18,77 +18,57 @@ import (
...
+		dialProtocol = "tcp6"
+		dialTarget = target
 	} else {
-		protocolProbeGauge.Set(4)
+		dialProtocol = "tcp4"
+		if port == "" {
```
Comment (brian-brazil) [inline tcp.go]: There has to be a port for TCP, so you can just use net.JoinHostPort.

**Comment by Conorbro** [comment]
@brian-brazil

**Comment by Conorbro** [comment]
@brian-brazil

File: main.go [hunk 46]
```diff
@@ -53,6 +53,7 @@ type Module struct {
 }
 
 type HTTPProbe struct {
+	PreferredIPProtocol string `yaml:"preferred_ip_protocol"`
```
Comment (brian-brazil) [inline main.go]: Have you run `make format` ?

File: tcp.go [hunk 47]
```diff
@@ -18,77 +18,53 @@ import (
...
 
-	if dialProtocol == "tcp6" {
-		protocolProbeGauge.Set(6)
+	if ip.IP.To4() == nil {
+		dialProtocol = "tcp6"
+		dialTarget = target
 	} else {
-		protocolProbeGauge.Set(4)
+		dialProtocol = "tcp4"
+		dialTarget = net.JoinHostPort(ip.String(), port)
```
Comment (brian-brazil) [inline tcp.go]: This is going to be the same for v6

**Comment by Conorbro** [comment]
@brian-brazil

File: tcp.go [hunk 48]
```diff
@@ -18,77 +18,53 @@ import (
...
+	ip, err := chooseProtocol(module.TCP.PreferredIPProtocol, targetAddress, registry)
+	if err != nil {
+		return nil, err
 	}
 
-	if dialProtocol == "tcp6" {
-		protocolProbeGauge.Set(6)
+	if ip.IP.To4() == nil {
+		dialProtocol = "tcp6"
+		dialTarget = net.JoinHostPort(ip.String(), port)
```
Comment (brian-brazil) [inline tcp.go]: this code is the same in both branches, it can be factored out

**Comment by Conorbro** [comment]
@brian-brazil

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**